kahuna/public/js/services/api/leases.js

```diff
@@ -81,54 +58,73 @@ leaseService.factory('leaseService', [
             });
 
             return image
-                .perform('replace-leases', {body: updatedLeases})
-                .then((op) => {
-                    return {
-                        op,
-                        originalLeaseCount
-                    };
-                });
-        }).then((op) => {
-            pollLeases(images, op.originalLeaseCount);
+              .perform('replace-leases', {body: updatedLeases})
+              .then(() => {
+                  pollLeases(images);
+              });
         });
     }
 
-    function batchAdd(lease, originalLeases, images) {
-      const originalLeaseCount = originalLeases.leases.length;
-      return $q.all(images.map(image => add(image, lease)))
-        .then(pollLeases(images, originalLeaseCount));
+    function add(image, lease) {
+      const newLease = angular.copy(lease);
+      newLease.mediaId = image.data.id;
+
+      if (angular.isDefined(newLease.notes) && newLease.notes.trim().length === 0) {
+        newLease.notes = null;
+      }
+
+      return image.perform('add-lease', {body: newLease});
+    }
+
+    function batchAdd(lease, images) {
+      return $q.all(images.map(image => add(image, lease))).then(() => {
+        pollLeases(images);
+      });
     }
 
     function canUserEdit(image){
       return editsService.canUserEdit(image);
     }
 
-    function deleteLease(lease, originalLeases, images) {
-      const originalLeaseCount = originalLeases.leases.length;
+    /**
+     * Delete a lease by uuid from a collection of images.
+     * This method does not support batch deletion, because a
+     * uuid will only ever match one lease.
+     */
+    function deleteLease(lease, images) {
       return getLeasesRoot().follow('leases', {id: lease.id}).delete()
-        .then(pollLeases(images, originalLeaseCount));
+        .then(() => pollLeases(images));
     }
 
     function getByMediaId(image) {
       return getLeasesRoot().follow('by-media-id', {id: image.data.id}).get();
     }
 
-    function pollLeases(images, originalLeaseCount){
+    function pollLeases(images) {
       apiPoll(() => {
-        return untilLeasesChange(images, originalLeaseCount);
+        return untilLeasesChange(images);
       });
     }
 
-    function untilLeasesChange(images, originalLeaseCount){
-      return $q.all(images.map((image) => image.get().then( (apiImage) => {
-        const apiLeases = imageAccessor.readLeases(apiImage);
-        if (apiLeases.leases.length !== originalLeaseCount) {
+    function untilLeasesChange(images) {
+      const imagesArray = images.toArray ? images.toArray() : images;
+      return $q.all(imagesArray.map(image => {
+        return image.get().then(apiImage => {
+          const apiLeases = imageAccessor.readLeases(apiImage);
+          const leases = imageAccessor.readLeases(image);
+          if (apiLeases.lastModified !== leases.lastModified) {
```

Comment:
minor: would requiring `apiLeases.lastModified` to be after `leases.lastModified` be better here?

Reply:
yes -- in fact, this had been written in one of my previous implementations. I'll make the update 👍